modules/fx/arrow-fx-kotlinx-coroutines/src/test/kotlin/arrow/integrations/kotlinx/CoroutinesIntegrationTest.kt

```diff
@@ -0,0 +1,166 @@
+          val promise = !Promise<Int>()
+          !effect {
+            scope.launch {
+              IO.cancelable<Unit> { promise.complete(i) }.suspendCancellable()
+            }
+          }
+          !effect { scope.cancel() }
+          !promise.get()
+        }.unsafeRunSync() == i
+      }
+    }
+
+    "suspendCancellable can cancel even for infinite asyncs" {
+      IO.async { cb: (Either<Throwable, Int>) -> Unit ->
+        val scope = TestCoroutineScope(Job() + TestCoroutineDispatcher())
+        scope.launch {
+          IO.async<Int> { Thread.sleep(5000) }
+            .onCancel(IO { cb(1.right()) })
+            .suspendCancellable().let {
+              cb(it.right())
+            }
+        }
+        IO(other) { Thread.sleep(500) }
+          .unsafeRunAsync { scope.cancel() }
+      }.unsafeRunTimed(2.seconds) shouldBe 1.some()
+    }
+
+    // --------------- unsafeRunScoped ---------------
+
+    "should rethrow exceptions within run block with unsafeRunScoped" {
+      forAll(Gen.throwable()) { e ->
+        try {
+          val scope = TestCoroutineScope(TestCoroutineDispatcher())
+          val ioa = IO<Int> { throw e }
+          ioa.unsafeRunScoped(scope) { either ->
+            either.fold({ throw it }, { fail("") })
+          }
+          fail("Should rethrow the exception")
+        } catch (throwable: Throwable) {
+          throwable == e
+        }
+      }
+    }
+
+    "unsafeRunScoped should cancel correctly" {
+      forAll(1, Gen.int()) { i ->
+        IO.async<Int> { cb ->
+          val scope = TestCoroutineScope(Job() + TestCoroutineDispatcher())
+          IO(all) { }
+            .flatMap { IO.async<Int> { cb -> Thread.sleep(200); cb(i.right()) } }
+            .unsafeRunScoped(scope) {
+              cb(it)
+            }
+          IO(other) { }
+            .unsafeRunAsync { scope.cancel() }
+        }.unsafeRunTimed(500.milliseconds) == None
+      }
+    }
+
+    "unsafeRunScoped can cancel even for infinite asyncs" {
+      IO.async { cb: (Either<Throwable, Int>) -> Unit ->
+        val scope = TestCoroutineScope(Job() + TestCoroutineDispatcher())
+        IO(all) { -1 }
+          .flatMap { IO.async<Int> { Thread.sleep(5000); } }
+          .onCancel(IO { cb(1.right()) })
+          .unsafeRunScoped(scope) {
+            cb(it)
+          }
+        IO(other) { Thread.sleep(500) }
+          .unsafeRunAsync { scope.cancel() }
+      }.unsafeRunTimed(2.seconds) shouldBe 1.some()
+    }
+
+    "should complete when running a pure value with unsafeRunScoped" {
+      forAll(Gen.int()) { i ->
+        val scope = TestCoroutineScope(TestCoroutineDispatcher())
+        IO.async<Int> { cb ->
+          IO.just(i).unsafeRunScoped(scope) { either ->
+            either.fold({ fail("") }, { cb(it.right()) })
+          }
+        }.unsafeRunSync() == i
+      }
+    }
+  }
+}
+
+fun <A> IOOf<A>.onCancel(token: IOOf<Unit>): IO<A> =
```

Comment:
I am raising a PR where this is moved into `Bracket<F, E>` and enabled concretely for `IO` ;)

Reply:
PR Raised #2017